src/grt/src/fastroute/src/DataProc.h

```diff
@@ -53,73 +53,46 @@ using boost::multi_array;
 
 // global variables
 extern int XRANGE, YRANGE;
-extern int xGrid, yGrid, numGrids, numNets, invalidNets, vCapacity, hCapacity;
-extern short *vCapacity3D, *hCapacity3D;
+extern int xGrid, yGrid, vCapacity, hCapacity;
 extern float vCapacity_lb, hCapacity_lb, vCapacity_ub, hCapacity_ub;
-extern int layerOrientation;
-extern float alpha;
 extern int verbose;
-extern int overflowIterations;
-const int max_usage_multiplier = 40;
 
 extern int enlarge, costheight, ripup_threshold;
-extern int MaxDegree;
-extern int *MinWidth, *MinSpacing, *ViaSpacing;
 extern int xcorner, ycorner, wTile, hTile, ahTH;
 
 extern int
     numValidNets;  // # nets need to be routed (having pins in different grids)
 extern int numLayers;
-extern int totalNumSeg;    // total # segments
 extern int totalOverflow;  // total # overflow
-extern int mazeThreshold;  // the wirelen threshold to do maze routing
 extern FrNet** nets;
-extern Edge *h_edges, *v_edges;
+extern std::vector<Edge> h_edges;
+extern std::vector<Edge> v_edges;
 
 extern multi_array<float, 2> d1;
 extern multi_array<float, 2> d2;
 
-extern Bool** HV;
-extern Bool** hyperV;
-extern Bool** hyperH;
+extern bool** HV;
+extern bool** hyperV;
+extern bool** hyperH;
 extern int** corrEdge;
-extern int SLOPE;
-
-// coefficient
-extern float LB;
-extern float UB;
-extern int THRESH_M;
-extern float LOGIS_COF;
-extern int ENLARGE;
-extern int STEP;
-extern int COSHEIGHT;
-extern int STOP;
-extern int L;
-extern int VCA;
-extern int VIA, slope, max_adj;
-extern char benchFile[STRINGLEN];
-
-extern Segment* seglist;
-extern int* seglistIndex;  // the index for the segments for each net
-extern int* seglistCnt;    // the number of segements for each net
-
-extern Tree* trees;      // the tree topologies
+
+extern std::vector<Segment> seglist;
+extern std::vector<int> seglistIndex;  // the index for the segments for each net
+extern std::vector<int> seglistCnt;    // the number of segements for each net
+
 extern StTree* sttrees;  // the Steiner trees
-extern DTYPE** gxs;      // the copy of xs for nets, used for second FLUTE
-extern DTYPE** gys;      // the copy of xs for nets, used for second FLUTE
-extern DTYPE**
-    gs;  // the copy of vertical sequence for nets, used for second FLUTE
+extern std::vector<std::vector<DTYPE>> gxs;        // the copy of xs for nets, used for second FLUTE
+extern std::vector<std::vector<DTYPE>> gys;        // the copy of xs for nets, used for second FLUTE
+extern std::vector<std::vector<DTYPE>> gs;  // the copy of vertical sequence for nets, used for second FLUTE
```

Comment:
For 2d arrays I've used multi_array in various places (eg d1, d2).   Would that work here?  Nested vectors involve extra memory allocations and produce less memory locality.

Reply:
The first dimension of these vectors has a fixed value equal to the number of nets to be routed. In FastRoute.cpp: `gxs.resize(numValidNets);`
The second dimension is the number of nodes of the specific net and is defined in another part of the code, in RSMT.cpp: `gxs[netID].resize(d);`.
I wasn't able to make it work with multi_array.